### PR TITLE
Allow passing of inputRef to RAC Checkbox/Switch

### PR DESCRIPTION
**Pull request by yihuiliao**
Opening to start a discussion about how we would like to handle inputRef since we ran into some issues with 🌈  Checkbox/Switch.

In RAC, inputRef is constructed within the component so that in our 🌈 implementation of Checkbox/Switch we have no way of passing an inputRef down to the input. We can't use `Context` because Checkbox/Switch don't use our `<Input>` component. 

If this is something we want, I can make adjustments to the ref support PR in 🌈 .

## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/WAI/ARIA/apg/)

### CODE REVIEW

File: packages/react-aria-components/src/Checkbox.tsx [hunk 1]
```diff
@@ -159,10 +159,14 @@ function CheckboxGroup(props: CheckboxGroupProps, ref: ForwardedRef<HTMLDivEleme
 
 export const CheckboxContext = createContext<ContextValue<CheckboxProps, HTMLLabelElement>>(null);
 
-function Checkbox(props: CheckboxProps, ref: ForwardedRef<HTMLLabelElement>) {
+function Checkbox(props: CheckboxProps & {inputRef?: RefObject<HTMLInputElement>}, ref: ForwardedRef<HTMLLabelElement>) {
   [props, ref] = useContextProps(props, ref, CheckboxContext);
-  let inputRef = useRef<HTMLInputElement>(null);
+  let {
+    inputRef: userInputRef
```
Comment (snowystinger) [inline packages/react-aria-components/src/Checkbox.tsx:165]: this won't merge the inputRef from a user provided prop vs the context (I think useContextProps will only merge the ref in the second argument)
we could move this before the `useContextProps` call, then we'd be getting the user provided one, and then we could merge that with the one from the context
```
let {inputRef: userProvidedInputRef, ...otherProps} = props;
[props, ref] = useContextProps(otherProps, ref, CheckboxContext);
// now merge the refs userProvidedInputRef and props.inputRef
```

another option we have is to not expose an inputRef and instead useImperative to add focus/select to the forwardedRef
but I think the types for that would be more difficult

File: packages/react-aria-components/test/Checkbox.test.js [hunk 2]
```diff
@@ -225,4 +225,10 @@ describe('Checkbox', () => {
     let {getByRole} = render(<Checkbox ref={ref}>Test</Checkbox>);
     expect(ref.current).toBe(getByRole('checkbox').closest('.react-aria-Checkbox'));
   });
+
+  it('should support input ref', () => {
```
Comment (snowystinger) [inline packages/react-aria-components/test/Checkbox.test.js:229]: can you add to the tests a version where a context inputRef is also passed?

File: packages/@react-aria/utils/src/mergeRefs.ts [hunk 3]
```diff
@@ -10,12 +10,12 @@
...
  */
 
-import {ForwardedRef} from 'react';
+import {ForwardedRef, MutableRefObject} from 'react';
 
 /**
  * Merges multiple refs into one. Works with either callback or object refs.
  */
-export function mergeRefs<T>(...refs: ForwardedRef<T>[]): ForwardedRef<T> {
+export function mergeRefs<T>(...refs: Array<ForwardedRef<T> | MutableRefObject<T>>): ForwardedRef<T> {
```
Comment (yihuiliao) [inline packages/@react-aria/utils/src/mergeRefs.ts:18]: had to update the typing for this but not super sure about this change

File: packages/react-aria-components/src/Switch.tsx [hunk 4]
```diff
@@ -62,9 +62,15 @@ export interface SwitchRenderProps {
 
 export const SwitchContext = createContext<ContextValue<SwitchProps, HTMLLabelElement>>(null);
 
-function Switch(props: SwitchProps, ref: ForwardedRef<HTMLLabelElement>) {
-  [props, ref] = useContextProps(props, ref, SwitchContext);
-  let inputRef = useRef<HTMLInputElement>(null);
+function Switch(props: SwitchProps & {inputRef?: MutableRefObject<HTMLInputElement>}, ref: ForwardedRef<HTMLLabelElement>) {
```
Comment (snowystinger) [inline packages/react-aria-components/src/Switch.tsx:65]: can we move the & added bit to the SwitchProps? or why is it here?

File: packages/@react-aria/utils/src/mergeRefs.ts (hunk 3 above)
Comment (snowystinger) [reply packages/@react-aria/utils/src/mergeRefs.ts:18]: react-aria/utils is in TS strict, so we're pretty well protected here
that said, I think it's the right change

File: packages/react-aria-components/src/Checkbox.tsx [hunk 5]
```diff
@@ -160,9 +165,15 @@ function CheckboxGroup(props: CheckboxGroupProps, ref: ForwardedRef<HTMLDivEleme
...
-  let inputRef = useRef<HTMLInputElement>(null);
+  let {
+    inputRef: userProvidedInputRef = null,
+    ...otherProps
+  } = props;
+  [props, ref] = useContextProps(otherProps, ref, CheckboxContext);
   let groupState = useContext(CheckboxGroupStateContext);
+  let defaultInputRef = useRef<HTMLInputElement>(null);
+  let userInputRef = useObjectRef(mergeRefs(userProvidedInputRef, props.inputRef !== undefined ? props.inputRef : null));
+  let inputRef = userInputRef || defaultInputRef;
```
Comment (LFDanLu) [inline packages/react-aria-components/src/Checkbox.tsx:176]: Small nit, but I believe useObjectRef always returns a value so you can get rid of `defaultInputRef` I believe.

```suggestion
  let inputRef = useObjectRef(mergeRefs(userProvidedInputRef, props.inputRef));
```

File: packages/react-aria-components/src/Switch.tsx [hunk 6]
```diff
@@ -63,8 +68,14 @@ export interface SwitchRenderProps {
...
-  [props, ref] = useContextProps(props, ref, SwitchContext);
-  let inputRef = useRef<HTMLInputElement>(null);
+  let {
+    inputRef: userProvidedInputRef = null,
+    ...otherProps
+  } = props;
+  [props, ref] = useContextProps(otherProps, ref, SwitchContext);
+  let defaultInputRef = useRef<HTMLInputElement>(null);
+  let userInputRef = useObjectRef(mergeRefs(userProvidedInputRef, props.inputRef !== undefined ? props.inputRef : null));
+  let inputRef = userInputRef || defaultInputRef;
```
Comment (LFDanLu) [inline packages/react-aria-components/src/Switch.tsx:78]: Same as the above
```suggestion
  let inputRef = useObjectRef(mergeRefs(userProvidedInputRef, props.inputRef));
```

**Review by LFDanLu — COMMENTED** [review]
Just small code nit, but otherwise this looks fine to me. Thanks for coming up with the update!

File: packages/react-aria-components/src/Checkbox.tsx (hunk 5 above)
Comment (yihuiliao) [reply packages/react-aria-components/src/Checkbox.tsx:176]: ah yep i think you're right, will remove!

**Review by snowystinger — APPROVED** [review]

**Review by devongovett — APPROVED** [review]

**Review by reidbarber — APPROVED** [review]

**Pull request merged**